lib/classes/ConfigSchemaHandler/errorMessageBuilder.js

```diff
@@ -0,0 +1,171 @@
+    ajvError => ajvError.keyword === 'anyOf' && ajvError.schemaPath === ANY_OF_EVENT_SCHEMA_PATH
+  );
+
+  // Alternatively, if there is a deeper error, then it's better to make it parent
+  // because it actually the reason why other errors are generated by AJV
+  const deeperAnyOfErrorIndex = ajvErrors.findIndex(ajvError =>
+    DEEPER_ANY_OF_SUB_ERROR_REGEX.test(ajvError.schemaPath)
+  );
+
+  const finalIndex = deeperAnyOfErrorIndex >= 0 ? deeperAnyOfErrorIndex : anyOfErrorIndex;
+
+  const result = [ajvErrors[finalIndex]];
+  result[0].params.suberrors = [];
+  ajvErrors.forEach((ajvError, index) => {
+    const isAnyOfSubError = ANY_OF_SUB_ERROR_REGEX.test(ajvError.schemaPath);
+
+    if (deeperAnyOfErrorIndex >= 0) {
+      if (isAnyOfSubError || index === anyOfErrorIndex) {
+        result[0].params.suberrors.push(ajvError);
+      } else if (index !== deeperAnyOfErrorIndex) {
+        result.push(ajvError);
+      }
+    }
+
+    if (deeperAnyOfErrorIndex < 0) {
+      if (isAnyOfSubError) {
+        result[0].params.suberrors.push(ajvError);
+      } else if (index !== anyOfErrorIndex) {
+        result.push(ajvError);
+      }
+    }
+  });
+
+  return result;
+};
+
+/*
+ * For error object structure, see https://github.com/ajv-validator/ajv/#error-objects
+ */
+const buildErrorMessages = (ajvErrors, userConfig = {}) => {
+  let errors = ajvErrors;
+
+  errors = structureAjvErrors(errors);
+
+  errors = addUserFriendlyMessage(errors, userConfig);
+
+  errors = errors.sort((a, b) => b.dataPath.length - a.dataPath.length);
+  if (!errors.length) throw new Error('Validation errors array cannot be emptry');
+
+  return errors.map(error => {
+    if (error.friendlyMessage) {
+      return error.friendlyMessage;
+    }
+    return buildDefaultErrorMessage(error);
+  });
+};
+
+const buildDefaultErrorMessage = ajvError => {
+  switch (ajvError.keyword) {
+    case 'additionalProperties': {
+      let additionalProperty;
+      if (ajvError.params && ajvError.params.additionalProperty) {
+        additionalProperty = ajvError.params.additionalProperty;
+      }
+      return `${ajvError.dataPath.slice(1)} ${ajvError.message} (received '${additionalProperty}')`;
+    }
+    default:
+      return `${ajvError.dataPath.slice(1)} ${ajvError.message}`;
+  }
+};
+
+const addUserFriendlyMessage = (errors, userConfig) => {
+  return errors.map(error => {
+    let friendlyMessage;
+
+    // regex matches '.functions['xxx'].events[x].xxx'
+    const isEventParam = /^\.functions\['[\w-_]+'\]\.events\[[0-9]+\]\.[\w]+$/.test(error.dataPath);
+    if (error.params && error.params.additionalProperty && isEventParam) {
+      const eventName = error.dataPath.split('.')[3];
+      friendlyMessage = `Unsupported parameter '${error.params.additionalProperty}' for '${eventName}' event`;
+    }
+
+    // regex matches '.functions['xxx'].events[x]'
+    const isFunctionEvent = /^\.functions\['[\w-_]+'\]\.events\[[0-9]+\]$/.test(error.dataPath);
```

Comment:
I wonder if there's an easy way to convert all paths as: `functions['function'].events[0]` (and I don't mean just for `functions`), into `functions.function.events[0]`, as this feels more natural.

I guess we may refactor it with help of simple regex (replace `\['([a-zA-Z_0-9-]+)'\]` with `$1`

Reply:
Good idea. I used a simpler regex `/\['(.+)'\]/g` though, hoping this is fine too

Reply:
I was thinking of replacing just those names which can literally be written as property names in JS.

Theoretically any string can be a property name, So removing `['` wrap for all is risky and in some corner case scenarios may make path really confusing.

That's why I proposed simple regex that will just rewrite those obvious literal names

Reply:
Thanks for the clarification, I've adjusted it in https://github.com/serverless/serverless/pull/7335/commits/4abac84afd74835efced7a75baaafe68d227e32f